File: generator.py
from Logics import logics

class Generator:
    """
        # Color Pallate Generator

        ### This is an Smart Color Pallate Generator for web developers and designer.
        generate full color pallate scheme by asking only one Main Color in `Hex Code` 
  
    """
    def __init__(self,hex) -> None:
        self.hex = hex
        logic = logics()
        self.rgb = logic.hextorgb(self.hex)
        self.hsv = logic.rgbtohsv(self.rgb[0],self.rgb[1],self.rgb[2])
        # variable for calculating shades. s for saturation and v for value
        self.__s19 = 22
        self.__s37 = 12
        self.__s26 = 17
        self.__s48 = 7

        self.__v19 = 30
        self.__v37 = 15
        self.__v26 = 22
        self.__v48 = 10

        # Creating Colors Default Hue amount
        self.__red = 1
        self.__orange = 37
        self.__blue = 210
        self.__green = 125

        self.__SupportingColorDict = {"red":{},"blue":{},"green":{},"orange":{}}
        

    def mainColor(self) -> dict:
        mainColorDict = {"hex":self.hex,"rgb":self.rgb,"hsv":self.hsv}
        return mainColorDict

    def NeutralColor(self) -> dict:
        Neutralcolor = {"100": "[0,0,94]",
                        "200": "[0,0,84]",
                        "300": "[0,0,74]",
                        "400": "[0,0,64]",
                        "500": "[0,0,54]",
                        "600": "[0,0,44]",
                        "700": "[0,0,34]",
                        "800": "[0,0,24]",
                        "900": "[0,0,10]"}
        return Neutralcolor

    def SupportingColor(self) -> dict:
        # Calculating Color Shades Dict and variable
        SaturationDict = {}
        ValueDict = {}
        # Saturation
        SaturationDict["100"] = self.hsv[1] - self.__s19
        SaturationDict["200"] = self.hsv[1] - self.__s26
        SaturationDict["300"] = self.hsv[1] - self.__s37
        SaturationDict["400"] = self.hsv[1] - self.__s48
        SaturationDict["500"] = self.hsv[1]
        SaturationDict["600"] = self.hsv[1] + self.__s26
        SaturationDict["700"] = self.hsv[1] + self.__s37
        SaturationDict["800"] = self.hsv[1] + self.__s48
        SaturationDict["900"] = self.hsv[1] + self.__s19

        # Value
        ValueDict["100"] = self.hsv[2] + self.__v19
        ValueDict["200"] = self.hsv[2] + self.__v26
        ValueDict["300"] = self.hsv[2] + self.__v37
        ValueDict["400"] = self.hsv[2] + self.__v48
        ValueDict["500"] = self.hsv[2]
        ValueDict["600"] = self.hsv[2] - self.__s26
        ValueDict["700"] = self.hsv[2] - self.__s37
        ValueDict["800"] = self.hsv[2] - self.__s48
        ValueDict["900"] = self.hsv[2] - self.__s19

        # applying Condition

        for k in SaturationDict.keys():
            if SaturationDict[k] < 10:
                SaturationDict[k] = 10
            elif SaturationDict[k] > 100:
                SaturationDict[k] = 100

        for k in ValueDict.keys():
            if ValueDict[k] < 15:
                ValueDict[k] = 15
            elif ValueDict[k] > 100:
                ValueDict[k] = 100
            
        # Generating Result 
        for cname in ["red","orange","green","blue"]:
            for k in ValueDict.keys():
                if cname == "red":
                    self.__SupportingColorDict[cname][k] = f"{self.__red,SaturationDict[k],ValueDict[k]}"
                elif cname == "orange":
                    self.__SupportingColorDict[cname][k] = f"{self.__orange,SaturationDict[k],ValueDict[k]}"
                elif cname == "green":
                    self.__SupportingColorDict[cname][k] = f"{self.__green,SaturationDict[k],ValueDict[k]}"
                else:
                    self.__SupportingColorDict[cname][k] = f"{self.__blue,SaturationDict[k],ValueDict[k]}"

        return self.__SupportingColorDict

File: JavaScript/generator.js
import {hextorgb,rgbtohsv} from "./logics.js";

class Generator{
    constructor(hex){
        this.hex = hex;
        this.rgb = hextorgb(this.hex);
        this.hsv = rgbtohsv(this.rgb[0],this.rgb[1],this.rgb[2]);

        // Variable for Calculating Shades, s for saturation and v for value

        this.s19 = 22;
        this.s37 = 12;
        this.s26 = 17;
        this.s48 = 7;

        this.v19 = 22;
        this.v37 = 12;
        this.v26 = 17;
        this.v48 = 7;

        // creating color defualt Hue Amount

        this.red = 1;
        this.orange = 37;
        this.blue = 210;
        this.green = 125;

        this.supportingColorObj = {"red":{},"blue":{},"green":{},"orange":{}};

    }

    mainColor(){
        const maincolorobj = {"hex":this.hex,"rgb":this.rgb,"hsv":this.hsv};
        return maincolorobj;
    }

    neutralColor(){
        const neutralcolorobj = {"100": "[0,0,94]",
        "200": "[0,0,84]",
        "300": "[0,0,74]",
        "400": "[0,0,64]",
        "500": "[0,0,54]",
        "600": "[0,0,44]",
        "700": "[0,0,34]",
        "800": "[0,0,24]",
        "900": "[0,0,10]"};

        return neutralcolorobj;
    }

    supportingColor(){
        // creating objects to handle output sobj for saturation and vobj for value
        let sobj = {};
        let vobj = {};

        // Saturation 
        sobj["100"] = this.hsv[1] - this.s19;
        sobj["200"] = this.hsv[1] - this.s26;
        sobj["300"] = this.hsv[1] - this.s37;
        sobj["400"] = this.hsv[1] - this.s48;
        sobj["500"] = this.hsv[1];
        sobj["600"] = this.hsv[1] + this.s26;
        sobj["700"] = this.hsv[1] + this.s37;
        sobj["800"] = this.hsv[1] + this.s48;
        sobj["900"] = this.hsv[1] + this.s19;

        // Value
        vobj["100"] = this.hsv[2] + this.s19;
        vobj["200"] = this.hsv[2] + this.s26;
        vobj["300"] = this.hsv[2] + this.s37;
        vobj["400"] = this.hsv[2] + this.s48;
        vobj["500"] = this.hsv[2];
        vobj["600"] = this.hsv[2] - this.s26;
        vobj["700"] = this.hsv[2] - this.s37;
        vobj["800"] = this.hsv[2] - this.s48;
        vobj["900"] = this.hsv[2] - this.s19;

        // Generating key Variable

        let objKey = Object.keys(sobj);
        

        // Looping through keys to CHeck conditions

        objKey.forEach(ele => {
            if(sobj[ele] < 10){
                sobj[ele] = 10;
            }else if(sobj[ele] > 100){
                sobj[ele] = 100;
            }
            if(vobj[ele] < 15){
                vobj[ele] = 15;
            }else if(vobj[ele] > 100){
                vobj[ele] = 100;
            }
        });

        // creating Color Name Variable
        // self.__SupportingColorDict[cname][k] = f"{self.__red,SaturationDict[k],ValueDict[k]}"

        let cname = ["red","green","blue","orange"];

        cname.forEach(color => {
            objKey.forEach(ind => {
                if (color === "red"){
                    this.supportingColorObj[color][ind] = [this.red,sobj[ind],vobj[ind]];
                }else if(color === "blue"){
                    this.supportingColorObj[color][ind] = [this.blue,sobj[ind],vobj[ind]];
                }else if(color === "green"){
                    this.supportingColorObj[color][ind] = [this.green,sobj[ind],vobj[ind]];
                }else{
                    this.supportingColorObj[color][ind] = [this.orange,sobj[ind],vobj[ind]];
                }
            });
        });

        return this.supportingColorObj

    }

}




